File: tests/runtime/results/leaf.rs
include!(env!("BINDINGS"));

use exports::test::results::test as imports;

pub struct Component;

export!(Component);

impl exports::test::results::test::Guest for Component {
    fn string_error(a: f32) -> Result<f32, String> {
        if a == 0.0 {
            Err("zero".to_owned())
        } else {
            Ok(a)
        }
    }

    fn enum_error(a: f32) -> Result<f32, imports::E> {
        if a == 0.0 {
            Err(imports::E::A)
        } else {
            Ok(a)
        }
    }

    fn record_error(a: f32) -> Result<f32, imports::E2> {
        if a == 0.0 {
            Err(imports::E2 {
                line: 420,
                column: 0,
            })
        } else if a == 1.0 {
            Err(imports::E2 {
                line: 77,
                column: 2,
            })
        } else {
            Ok(a)
        }
    }

    fn variant_error(a: f32) -> Result<f32, imports::E3> {
        if a == 0.0 {
            Err(imports::E3::E2(imports::E2 {
                line: 420,
                column: 0,
            }))
        } else if a == 1.0 {
            Err(imports::E3::E1(imports::E::B))
        } else if a == 2.0 {
            Err(imports::E3::E1(imports::E::C))
        } else {
            Ok(a)
        }
    }

    fn empty_error(a: u32) -> Result<u32, ()> {
        if a == 0 {
            Err(())
        } else if a == 1 {
            Ok(42)
        } else {
            Ok(a)
        }
    }

    fn double_error(a: u32) -> Result<Result<(), String>, String> {
        if a == 0 {
            Ok(Ok(()))
        } else if a == 1 {
            Ok(Err("one".into()))
        } else {
            Err("two".into())
        }
    }
}

File: tests/runtime/results/leaf.cpp
#include <assert.h>
#include <leaf_cpp.h>

template <class T>
bool equal(T const&a, T const&b) {
    return a==b;
}

namespace test_exports = ::exports::test::results::test;

std::expected<float, wit::string> test_exports::StringError(float a) {
    if (a==0.0) {
        return std::unexpected(wit::string::from_view("zero"));
    }
    else {
        return a;
    }
}

std::expected<float, test_exports::E> test_exports::EnumError(float a) {
    if (a==0.0) {
        return std::unexpected(test_exports::E::kA);
    }
    else {
        return a;
    }
}

std::expected<float, test_exports::E2> test_exports::RecordError(float a) {
    if (a==0.0) {
        return std::unexpected(test_exports::E2{420, 0});
    }
    else if (a==1.0) {
        return std::unexpected(test_exports::E2{77, 2});
    }
    else {
        return a;
    }
}

std::expected<float, test_exports::E3> test_exports::VariantError(float a) {
    if (a==0.0) {
        return std::unexpected(test_exports::E3{test_exports::E3::E2{test_exports::E2{420, 0}}});
    }
    else if (a==1.0) {
        return std::unexpected(test_exports::E3{test_exports::E3::E1{test_exports::E::kB}});
    }
    else if (a==2.0) {
        return std::unexpected(test_exports::E3{test_exports::E3::E1{test_exports::E::kC}});
    }
    else {
        return a;
    }
}

std::expected<uint32_t, wit::Void> test_exports::EmptyError(uint32_t a) {
    if (a==0) {
        return std::unexpected(wit::Void{});
    }
    else if (a==1) {
        return 42;
    }
    else {
        return a;
    }
}

std::expected<std::expected<void, wit::string>, wit::string> test_exports::DoubleError(uint32_t a) {
    if (a==0) {
        return std::expected<void, wit::string>();
    }
    else if (a==1) {
        return std::expected<void, wit::string>(std::unexpected(wit::string::from_view("one")));
    }
    else {
        return std::unexpected(wit::string::from_view("two"));
    }
}
